packages/monitor-v2/src/monitor-og/oSnapAutomation.ts

```diff
@@ -0,0 +1,393 @@
+  );
+
+  // Get all finalized basic safeSnap proposals for supported spaces.
+  const snapshotProposals = (
+    await Promise.all(
+      supportedSpaces.map(async (space) => getSnapshotProposals(space, params.graphqlEndpoint, params.retryOptions))
+    )
+  ).flat();
+
+  // Expand Snapshot proposals to include only one safe per proposal.
+  const expandedProposals: SnapshotProposalExpanded[] = snapshotProposals.flatMap((proposal) => {
+    const { plugins, ...clonedObject } = proposal;
+    return proposal.plugins.safeSnap.safes.map((safe) => ({ ...clonedObject, safe }));
+  });
+
+  // Return only proposals from supported safes.
+  return expandedProposals.filter((proposal) => isSafeSupported(proposal.safe, supportedModules, params.chainId));
+};
+
+// Get all proposals on supported oSnap modules that have not been discarded. Discards are most likely due to disputes,
+// but can also occur on OOv3 upgrades.
+const getUndiscardedProposals = async (
+  supportedModules: SupportedModules,
+  params: MonitoringParams
+): Promise<Array<TransactionsProposedEvent>> => {
+  // Get all proposals for all supported modules.
+  const allProposals = (
+    await Promise.all(
+      Object.keys(supportedModules).map(async (ogAddress) => {
+        const og = await getOgByAddress(params, ogAddress);
+        return runQueryFilter<TransactionsProposedEvent>(og, og.filters.TransactionsProposed(), {
+          start: 0,
+          end: params.blockRange.end,
+        });
+      })
+    )
+  ).flat();
+
+  // Get all deleted proposals for all supported modules.
+  const deletedProposals = (
+    await Promise.all(
+      Object.keys(supportedModules).map(async (ogAddress) => {
+        const og = await getOgByAddress(params, ogAddress);
+        return runQueryFilter<ProposalDeletedEvent>(og, og.filters.ProposalDeleted(), {
+          start: 0,
+          end: params.blockRange.end,
+        });
+      })
+    )
+  ).flat();
+
+  // Filter out all proposals that have been deleted by matching assertionId. assertionId should be sufficient property
+  // for filtering as it is derived from module address, transaction content and assertion time among other factors.
+  const deletedAssertionIds = deletedProposals.map((deletedProposal) => deletedProposal.args.assertionId);
+  return allProposals.filter((proposal) => !deletedAssertionIds.includes(proposal.args.assertionId));
+};
+
+// Checks if a safeSnap safe from Snapshot proposal is supported by oSnap automation.
+const isSafeSupported = (safe: SafeSnapSafe, supportedModules: SupportedModules, chainId: number): boolean => {
+  for (const ogAddress in supportedModules) {
+    if (isMatchingSafe(safe, chainId, ogAddress)) return true;
+  }
+  return false;
+};
+
+// Filters out all Snapshot proposals that have been proposed on-chain. This is done by matching safe, explanation and
+// proposed transactions.
+const filterPotentialProposals = (
```

Comment:
Are explanations + transactions always unique? If a safe proposes the same txs and same explanation two times at different moments would this function filter the second proposal wrongly? Just checking, I'm not sure about this.

Reply:
This filter is used to narrow down potential proposals to post on-chain. So it would not make a difference if there are 1, 2 or even more matching on-chain proposals as the bot would not post this.